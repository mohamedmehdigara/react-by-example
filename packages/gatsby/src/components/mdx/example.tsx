import React, { FC, ReactNode, useCallback, useState } from 'react'
import styled from 'styled-components'
import { CodeSandboxIcon, ResetIcon } from '../icons'

const Render = styled.div`
  display: flex;
  padding: 1em;
  border: 1px solid var(--app-foreground);
  border-radius: var(--app-border-radius);
  align-items: center;
  justify-content: center;
`
const FlexRight = styled.div`
  display: flex;
  justify-content: flex-end;
`

const CodeSandboxLink = styled.a`
  display: flex;
`

const ResetButton = styled.button`
  display: flex;
  margin: 0;
  padding: 0;
  color: var(--app-foreground);
  background: none;
  border: none;
  cursor: pointer;

  :focus,
  :active {
    color: var(--app-cyan);
    outline: 2px solid var(--app-cyan);
  }
`

export interface ExampleProps {
  code?: ReactNode
  render?: ReactNode
  link?: string
}

export const Example: FC<ExampleProps> = ({ code, render, link }) => {
  const [resetKey, setResetKey] = useState(Date.now())
  const triggerReset = useCallback(() => {
    setResetKey(Date.now())
  }, [])
  return (
    <>
      {code}
      <Render key={resetKey}>
        <div>{render}</div>
      </Render>
      {link && (
        <FlexRight>
          <ResetButton
            type="button"
            title="reset example"
            onClick={triggerReset}
          >
            <ResetIcon />
          </ResetButton>
          <CodeSandboxLink
            target="_blank"
            rel="noreferrer"
            href={link}
            title="edit on CodeSandbox"
          >
            <CodeSandboxIcon />
          </CodeSandboxLink>
        </FlexRight>
      )}
    </>
  )
}
