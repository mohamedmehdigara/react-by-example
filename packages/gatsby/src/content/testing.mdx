---
title: Testing
---

## Testing

- [Arrange-Act-Assert][aaa] pattern works great with React

[aaa]: https://duckduckgo.com/?q=arrange+act+assert

```yaml {example}
type: test
project: testing-dummy
module: src/index.test.ts
```

### `GifFinder`

- let's test a component that uses state & effects to find gifs

```yaml {example}
project: testing-components
style:
  width: 100%
  height: 50em
```

### Testing `GifFinder`

> The more your tests resemble the way your software is used,
> the more confidence they can give you.
>
> -- [`@testing-library`]

- [`create-react-app`] comes with [`@testing-library`]
- works with [angular], [vue], [plain DOM][dom], etc...
- no selectors, [forcing] you to write accessible components
- alternatively, use [enzyme]

[`create-react-app`]: https://create-react-app.dev/
[`@testing-library`]: https://testing-library.com/
[angular]: https://testing-library.com/docs/angular-testing-library/intro
[vue]: https://testing-library.com/docs/vue-testing-library/intro
[dom]: https://testing-library.com/docs/dom-testing-library/intro
[forcing]: https://coderefinery.nz/2020/10/21/forcing-functions-in-software-development/
[enzyme]: https://github.com/enzymejs/enzyme

```yaml {example}
type: test
project: testing-components-test
```
