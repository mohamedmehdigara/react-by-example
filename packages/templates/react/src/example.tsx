import React, { FC, ReactElement } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default (): ReactElement => <Simple />
