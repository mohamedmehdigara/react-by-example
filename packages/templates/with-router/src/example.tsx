import React, { FC } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default <Simple />
