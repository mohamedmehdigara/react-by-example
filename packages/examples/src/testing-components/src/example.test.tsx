import {
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from '@testing-library/react'
import React from 'react'
import { GifFinder } from './example'
import { TenorApi } from './tenor-api'
import { promiseShim } from './testing/promise-shim'

jest.mock('./tenor-api', () => ({
  TenorApi: {
    search: jest.fn(() => Promise.resolve(['gif1', 'gif2', 'gif3'])),
  },
}))

describe('GifFinder', () => {
  let component: RenderResult
  let onFound: jest.Mock

  promiseShim()

  beforeEach(() => {
    onFound = jest.fn()
    component = render(<GifFinder onFound={onFound} />)
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  describe('when searching', () => {
    let queryInput: HTMLElement

    beforeEach(() => {
      queryInput = component.getByLabelText('find a gif')
      fireEvent.input(queryInput, { target: { value: 'react' } })
    })

    it('should start a search', () => {
      // eslint-disable-next-line @typescript-eslint/unbound-method
      expect(TenorApi.search).toHaveBeenCalledWith('react')

      // eslint-disable-next-line @typescript-eslint/unbound-method
      expect(TenorApi.search).toHaveBeenCalledTimes(1)
    })

    it('should render the gifs', () => {
      expect(component.container).toMatchSnapshot()
    })

    describe('when clicking a gif', () => {
      beforeEach(() => {
        const [, gif2] = component.getAllByLabelText('gif')
        fireEvent.click(gif2)
      })

      it('should call onFound', () => {
        expect(onFound).toHaveBeenCalledWith('gif2')
        expect(onFound).toHaveBeenCalledTimes(1)
      })
    })
  })
})
