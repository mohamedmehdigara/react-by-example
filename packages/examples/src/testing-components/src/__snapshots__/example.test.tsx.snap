// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifFinder should snapshot 1`] = `
<div>
  <section>
    <label>
      find a gif
      <input
        placeholder="query"
        value=""
      />
    </label>
  </section>
</div>
`;

exports[`GifFinder when searching should render the gifs 1`] = `
<div>
  <section>
    <label>
      find a gif
      <input
        placeholder="query"
        value="react"
      />
    </label>
    <div
      class="grid"
    >
      <button
        aria-label="gif"
        class="button"
        type="button"
      >
        <img
          alt="gif"
          class="image"
          src="gif1"
        />
      </button>
      <button
        aria-label="gif"
        class="button"
        type="button"
      >
        <img
          alt="gif"
          class="image"
          src="gif2"
        />
      </button>
      <button
        aria-label="gif"
        class="button"
        type="button"
      >
        <img
          alt="gif"
          class="image"
          src="gif3"
        />
      </button>
    </div>
  </section>
</div>
`;
